### PR TITLE
Flex time per day

### LINKED ISSUE #47: option to have a weekly flexi time instead of a global/monthly one
Or even a daily one

### PR DESCRIPTION
**Pull request by dirkmb**
This PR fixes #47 'weekly/daily flexi time'

- Added a new event type called 'FLEX' which overrides the default FLEX-Time for this day.
- FLEX-Time is saved for each week in the database, so if the default flex time is changed over the time, the old weeks keep their flex time (until you re import the data)

My editor removed all trailing white spaces, sorry for theses many changes. Since I'm not a Java programmer I'm not sure if theses spaces where intentional. If you want to keep them let me know.

### CODE REVIEW

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java [hunk 1]
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
 	public DayLine calulateOneDay(DateTime day, List<Event> eventsOfOneDay) {
 		DayLine ret = new DayLine();
 
+        boolean found_day_flex_time = false;
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java]:
```suggestion
        boolean foundDayFlexTime = false;
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java [hunk 2]
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
...
+        for (Event event : eventsOfOneDay) {
+            if (event.getType() == TypeEnum.FLEX.getValue()) {
+                DateTime flex_time = DateTimeUtil.stringToDateTime(event.getTime());
+                ret.getTimeFlexi().substract(flex_time.getHour(), flex_time.getMinute());
+                found_day_flex_time = true;
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java]:
```suggestion
                foundDayFlexTime = true;
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java [hunk 3]
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
 	public DayLine calulateOneDay(DateTime day, List<Event> eventsOfOneDay) {
 		DayLine ret = new DayLine();
 
+        boolean found_day_flex_time = false;
+        for (Event event : eventsOfOneDay) {
+            if (event.getType() == TypeEnum.FLEX.getValue()) {
+                DateTime flex_time = DateTimeUtil.stringToDateTime(event.getTime());
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java]:
```suggestion
                DateTime flexTime = DateTimeUtil.stringToDateTime(event.getTime());
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java [hunk 4]
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
...
+                ret.getTimeFlexi().substract(flex_time.getHour(), flex_time.getMinute());
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java]:
```suggestion
                ret.getTimeFlexi().substract(flexTime.getHour(), flexTime.getMinute());
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java [hunk 5]
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
...
+                found_day_flex_time = true;
+                break;
+            }
+        }
+
+        // get default flex time from settings
 		WeekDayEnum weekDay = WeekDayEnum.getByValue(day.getWeekDay());
 		if (Basics.getInstance().getPreferences().getBoolean(Key.ENABLE_FLEXI_TIME.getName(), false)
-			&& timerManager.isWorkDay(weekDay)) {
+			&& timerManager.isWorkDay(weekDay) && found_day_flex_time == false) {
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java]:
```suggestion
			&& timerManager.isWorkDay(weekDay) && foundDayFlexTime == false) {
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java [hunk 6]
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
...
 			// substract the "normal" work time for one day
 			int normalWorkTimeInMinutes = timerManager.getNormalWorkDurationFor(weekDay);
 			ret.getTimeFlexi().substract(0, normalWorkTimeInMinutes);
 		}
 
-		if (eventsOfOneDay == null || eventsOfOneDay.isEmpty()) {
+		if (eventsOfOneDay == null || eventsOfOneDay.isEmpty() || found_day_flex_time && eventsOfOneDay.size() == 1) {
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java]:
```suggestion
		if (eventsOfOneDay == null || eventsOfOneDay.isEmpty() || foundDayFlexTime && eventsOfOneDay.size() == 1) {
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java [hunk 7]
```diff
@@ -308,11 +315,51 @@ public TimeSum calculateTimeSum(DateTime date, PeriodEnum periodEnum) {
...
+		TimeSum ret = new TimeSum();
+		List<Event> events = null;
+
+		// make date the first day of this week
+		date = date.minusDays(date.getWeekDay() - 1);
+
+		for (WeekDayEnum weekDay : WeekDayEnum.values()) {
+			DateTime day = date.plusDays(weekDay.getValue() - 1);
+			events = dao.getEventsOnDay(day);
+			boolean found_flex_time = false;
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java]:
```suggestion
			boolean foundFlexTime = false;
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java [hunk 8]
```diff
@@ -308,11 +315,51 @@ public TimeSum calculateTimeSum(DateTime date, PeriodEnum periodEnum) {
...
+
+				// skip all events that are not flex events
+				if (!isFlexEvent(event)) {
+					continue;
+				}
+
+				DateTime flex_time = DateTimeUtil.stringToDateTime(event.getTime());
+				ret.add(eventTime.getHour(), eventTime.getMinute());
+				Logger.debug("calculating flexi time for {} {} event found: {}", day.toString(), weekDay.getValue(), flex_time.toString());
+				found_flex_time = true;
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java]:
```suggestion
				foundFlexTime = true;
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java [hunk 9]
```diff
@@ -308,11 +315,51 @@ public TimeSum calculateTimeSum(DateTime date, PeriodEnum periodEnum) {
...
+				break;  // there can only be one flext time per day
+			}
+			if (found_flex_time == false) {
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java]:
```suggestion
			if (foundFlexTime == false) {
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java [hunk 10]
```diff
@@ -308,11 +315,51 @@ public TimeSum calculateTimeSum(DateTime date, PeriodEnum periodEnum) {
...
+				// skip all events that are not flex events
+				if (!isFlexEvent(event)) {
+					continue;
+				}
+
+				DateTime flex_time = DateTimeUtil.stringToDateTime(event.getTime());
+				ret.add(eventTime.getHour(), eventTime.getMinute());
+				Logger.debug("calculating flexi time for {} {} event found: {}", day.toString(), weekDay.getValue(), flex_time.toString());
+				found_flex_time = true;
+				break;  // there can only be one flext time per day
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java]:
```suggestion
				break;  // there can only be one flexi time per day
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java [hunk 11]
```diff
@@ -604,13 +658,19 @@ public void updateWeekSum(final Week week) {
 			Logger.warn("NOT setting the sum of the week starting {} to {} minutes (the sum cannot be negative)", weekToUse.getStart(), minutes);
 			weekToUse.setSum(0);
 		}
+
+		TimeSum flexi_sum = calculateFlexTimeSum(DateTimeUtil.stringToDateTime(week.getStart()));
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java]:
```suggestion
		TimeSum flexiSum = calculateFlexTimeSum(DateTimeUtil.stringToDateTime(week.getStart()));
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java [hunk 12]
```diff
@@ -604,13 +658,19 @@ public void updateWeekSum(final Week week) {
...
+		int flexi_minutes = flexi_sum.getAsMinutes();
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java]:
```suggestion
		int flexiMinutes = flexiSum.getAsMinutes();
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java [hunk 13]
```diff
@@ -604,13 +658,19 @@ public void updateWeekSum(final Week week) {
...
+		Logger.info("updating the flexi time sum to {} minutes for the week beginning at {}", flexi_minutes, week.getStart());
+		weekToUse.setFlexi(flexi_minutes);
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java]:
```suggestion
		weekToUse.setFlexi(flexiMinutes);
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java [hunk 14]
```diff
@@ -604,13 +658,19 @@ public void updateWeekSum(final Week week) {
 			Logger.warn("NOT setting the sum of the week starting {} to {} minutes (the sum cannot be negative)", weekToUse.getStart(), minutes);
 			weekToUse.setSum(0);
 		}
+
+		TimeSum flexi_sum = calculateFlexTimeSum(DateTimeUtil.stringToDateTime(week.getStart()));
+		int flexi_minutes = flexi_sum.getAsMinutes();
+		Logger.info("updating the flexi time sum to {} minutes for the week beginning at {}", flexi_minutes, week.getStart());
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java]:
```suggestion
		Logger.info("updating the flexi time sum to {} minutes for the week beginning at {}", flexiMinutes, week.getStart());
```

**Comment by mathisdt** [comment]
Next steps: when the variable name issue is resolved, I will merge the PR and run the new dev version on my own device for a while before creating an official new version of the app. But I won't disregard it - it just may take some time as I do this on my (rare) spare time.

**Comment by dirkmb** [comment]
hi, thx for the fast and detailed review. I changed all variable names and updated the MR.
Sorry it took so long, like you I also had only very rare spare time.

**Comment by mathisdt** [comment]
I'm not sure I have understood what the new feature is doing.

1. I have the following situation without using a flex event. No work was tracked on Thursday so flexi time is reduced.
events-without-flex
overview-without-flex

2. Now I add a flex event on Thursday and have the following:
events-with-flex
overview-with-flex

As far as I can see, the flex event prohibits that flexi time is reduced on a day on which no work was tracked (Thursday in the above example). Is that what the flex event should do?

**Comment by dirkmb** [comment]
I guess the use of 'FLEX' as event name confused you, or it does not say what it is doing. This event defines the needed/wanted flex time for the given day. So instead of 8 hours you are supposed to work only 0 hours on Thursday.

You have already 9 hours on your flexi-time (overtime). After you configured the FLEX event for Thursday you needed to work 0 hours for this day. So you do have the previous 9 hours of flexi-time on you clock.
If you would work 1 hour on Thursday you would have 10 hours flexi-time.

Does this explain it?

**Pull request merged**